Add CofreeT

eviefp opened the pull request:
**Description of the change**

I'm looking for early feedback.

Here's my attempt at implementing `CofreeT`. I'm not very sure it's correct, so I would love some feedback. I've mostly followed the existing `FreeT` code and looked at the Haskell `CofreeT` signatures.

Fixes #14 

I will definitely add documentation, and follow the rest of the checklist as appropriate. I'm currently looking for early feedback on the implementation. If it would make it easier for you to provide feedback on a "complete" PR (i.e. including docs, tests, etc.), I would be happy to oblige.

---

**Checklist:**

- [x] Added the change to the changelog's "Unreleased" section with a link to this PR and your username
- [x] Linked any existing issues or proposals that this pull request should close
- [x] Updated or added relevant documentation in the README and/or documentation directory
- [ ] Added a test for the contribution (if applicable)

Commit:
Initial CofreeT implementation.

JordanMartinez commented:
I'm not familiar enough with CofreeT to comment on this. Overall, it looks like the initial PR is good. My guess is that we're not looking for the most performant implementation of this, or even the most complete implementation; merely, an implementation in the first place, correct?

eviefp commented:
Alright, thank you for the feedback. I'll proceed to add docs, tests, etc., as to make it ready for review/merge!

Commit:
Add example and documentation.

eviefp commented:
I have no idea how to write unit tests for `CofreeT` since all the functions seem to only have one implementation which type-checks. I think the only "interesting" choice in the module is actually the type itself.

Note that, despite me having added a file under `test/`, it's not actually a test: it's just an example.

I think this is now ready for review!

JordanMartinez commented:
LGTM. The docs helped me understand why I would use this, so thanks!

thomashoneyman commented:
I'll have time to review this certainly by the end of the week. Thanks for the PR!

eviefp commented:
I realised I might want to do one more thing: consider what other instances I could add to `CofreeT`. I think there's quite a few. I'll have them ready before the weekend :)

EDIT: nevermind, they are done :point_down:

Commit:
Add a bunch of instances.

thomashoneyman commented on src/Control/Monad/Cofree/Trans.purs:
```diff
@@ -26,6 +38,13 @@ cofreeT
+    -> CofreeT f m a
+cofreeT' t = CofreeT $ (\_ -> t)
+
+
```
```suggestion
```

thomashoneyman commented on CHANGELOG.md:
```diff
@@ -7,6 +7,7 @@ Notable changes to this project are documented in this file. The format is based
 Breaking changes (😱!!!):
 
 New features:
+- added the `CofreeT`, the cofree monad transformer in a new module (`Control.Monad.Cofree.Trans`)
```
I think this is technically a breaking change due to the addition of two dependencies (`free` and `tuples`). Though it's hard to imagine someone using `FreeT` without already having those dependencies brought in.

thomashoneyman replied:
```suggestion
- added the `CofreeT`, the cofree monad transformer in a new module (`Control.Monad.Cofree.Trans`) ([#26](https://github.com/purescript-contrib/purescript-freet/pull/26))
```

thomashoneyman commented on src/Control/Monad/Cofree/Trans.purs:
```diff
@@ -0,0 +1,158 @@
+tail :: forall f m a. Functor m => CofreeT f m a -> m (f (CofreeT f m a))
+tail = map T.snd <<< unCofreeT
+
+-- Note: This cannot be automatically derived because 'a' also appers in the
```
```suggestion
-- Note: This cannot be automatically derived because 'a' also appears in the
```

thomashoneyman commented on src/Control/Monad/Cofree/Trans.purs:
```diff
@@ -0,0 +1,158 @@
+
+instance comonadAskCofreeT :: (Functor f, ComonadAsk e m) => ComonadAsk e (CofreeT f m) where
+  ask = ask <<< tail
+
```
```suggestion
```

thomashoneyman commented on src/Control/Monad/Cofree/Trans.purs:
```diff
@@ -0,0 +1,158 @@
+instance ordCofreeT :: Ord (m (Tuple a (f (CofreeT f m a)))) => Ord (CofreeT f m a) where
+  compare = compare `on` unCofreeT
+
+
```
```suggestion
```

thomashoneyman commented on src/Control/Monad/Cofree/Trans.purs:
```diff
@@ -0,0 +1,158 @@
+
+
+-- | Unpack `CofreeT` into the inner computation.
+unCofreeT :: forall f m a. CofreeT f m a -> m (Tuple a (f (CofreeT f m a)))
```
`FreeT` uses `runFreeT` for this, so we should probably use the same naming convention here (`runCofreeT`).

thomashoneyman commented on src/Control/Monad/Cofree/Trans.purs:
```diff
@@ -0,0 +1,158 @@
+    -> CofreeT f m a
+cofreeT = CofreeT
+
+cofreeT'
```
Do you mind adding a comment? Perhaps even something simple like:

```
-- | Construct a `CofreeT` from a computation with an annotation 'a'.
```

thomashoneyman reviewed (commented):
I'm not particularly familiar with `Cofree`, so it would be nice to have another look from @garyb (if you have a moment or suggestions for anyone else who should verify this). That said, it looks pretty good to me! Just a couple of suggestions.

thomashoneyman commented:
There is also this implementation which relies on `Lazy` (though yours via `Unit -> ...` accomplishes the same thing), which you may find interesting to compare:

https://github.com/arthurxavierx/purescript-comonad-ui-todos/blob/master/src/Control/Comonad/Cofree/Trans.purs

Commit:
Address Thomas' review comments.

Commit:
Remove extra vertical whitespace.

eviefp commented:
I addressed all review comments.

About the `Lazy` vs `Unit ->` bit, the only reason I would keep it as-is is because we're not already depending on `lazy`. But if you think it's better to add that dependency, it's a trivial change. I haven't used `lazy` much (I even forgot it exists) -- that's the only reason I didn't implement it like that from the beginning.

Commit:
Fix markdown list.

thomashoneyman commented:
No need to introduce it, although since we’re already bringing in new dependencies it wouldn’t be that big of a deal.

I’d like to let this sit for a week or so for others to weigh in, but if no one does by then I will go ahead and merge because it appears correct to me (with the caveat that I’m not particularly familiar with this construct).

paf31 commented on src/Control/Monad/Cofree/Trans.purs:
```diff
@@ -0,0 +1,156 @@
+import Effect.Aff.Class (class MonadAff, liftAff)
+import Effect.Class (class MonadEffect, liftEffect)
+
+-- | The cofree monad transformer for the functor 'f'.
```
_monad_ transformer? Do you mean comonad transformer? I know it's also a monad transformer, but it's hardly the main use case.

paf31 replied:
Also, in that light, it's customary to name the argument `w`, not `m`.

paf31 replied:
Also because in a category theoretic sense, it is the cofree comonad transformer in the sense that it's an adjoint to the forgetful functor to natural transformations.

paf31 commented on src/Control/Monad/Cofree/Trans.purs:
```diff
@@ -0,0 +1,156 @@
+-- | Usually, you would use `CofreeT` to annotate an existing AST with
+-- | metadata such as source locations, file names, etc.
+
+module Control.Monad.Cofree.Trans where
```
Same here - odd choice of module name?

thomashoneyman replied:
Whoops, missed that in review. I agree, the module name should be changed.

paf31 commented on src/Control/Monad/Cofree/Trans.purs:
```diff
@@ -0,0 +1,156 @@
+instance comonadCofreeT :: (Comonad m, Functor f) => Comonad (CofreeT f m) where
+  extract = extract <<< head
+
+instance eqCofreeT :: Eq (m (Tuple a (f (CofreeT f m a)))) => Eq (CofreeT f m a) where
```
This won't be guaranteed to terminate in most actual use cases. Is there a reason for adding it?

eviefp replied:
Is it because we expect `CofreeT`s to generally be infinite? The main reason I added it is because this instance exists in Haskell as well, but I see your point. I'm not sure this would get a lot of (legitimate) use. Maybe it's better to remove and consider re-adding if we get requests/issues for real use cases.

garyb replied:
Would it be possible to use `Eq1 m` instead of this constraint, if we keep it?

eviefp replied:
I removed both `Eq` and `Ord` for now. It might be possible to use `Eq1` though, so I'll keep that in mind if we end up implementing it eventually.

Commit:
Move to Comonad, rename tyvars, and remove Eq/Ord.

Commit:
Fix test example.

eviefp commented:
I think I addressed all review comments. Thank you for taking your time to review my PR!
